filters/tags_filter.go

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package filters
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+)
+
+// SortedTagIterator iterates over a set of tag names and values
+// sorted by tag names in ascending order
+type SortedTagIterator interface {
+	// Next returns true if there are more tag names and values
+	Next() bool
+
+	// Current returns the current tag name and value
+	Current() (string, string)
+
+	// Err returns any errors encountered
+	Err() error
+
+	// Close closes the iterator
+	Close()
+}
+
+// NewSortedTagIteratorFn creates a tag iterator given an id
+type NewSortedTagIteratorFn func(id string) SortedTagIterator
+
+// tagFilter is a filter associated with a given tag
+type tagFilter struct {
+	name        string
+	valueFilter Filter
+}
+
+func (f *tagFilter) String() string {
+	return fmt.Sprintf("%s:%s", f.name, f.valueFilter.String())
+}
+
+type tagFiltersByNameAsc []tagFilter
+
+func (tn tagFiltersByNameAsc) Len() int           { return len(tn) }
+func (tn tagFiltersByNameAsc) Swap(i, j int)      { tn[i], tn[j] = tn[j], tn[i] }
+func (tn tagFiltersByNameAsc) Less(i, j int) bool { return tn[i].name < tn[j].name }
+
+// tagsFilter contains a list of tag filters.
+type tagsFilter struct {
+	filters []tagFilter
+	iterFn  NewSortedTagIteratorFn
+	op      LogicalOp
```

Comment:
This was mostly in preparation for combinations of filters (ie eventually if you want to create a rule like (tag1=val1 || tag2=val1) && tag3=val3, you can do this with one rule instead of multiple. Also, the OR semantic is required in other parts of the stack, so figured I'd plumb it through here. Doesn't really hurt performance for the base case, the conjunction case is checked first most times. Happy to turn this into an enum too if you're worried about mem cost, string was just nice for printing.